File: src/main/kotlin/ca/yakov/aoc/aoc2016/day16.kt
package ca.yakov.aoc.aoc2016

import ca.yakov.aoc.Solution

fun run16(input: String): Solution {
    val inputTrimmed = input.trim()
    return Solution(
        checkMe(inputTrimmed, 272),
        checkMe(inputTrimmed, 35651584)
    )
}

private fun checkMe(input: String, length: Int): ULong {
    var a = input
    while (a.length < length) {
        val b = a.reversed().map { if (it == '0') '1' else '0' }.joinToString("")
        a += "0$b"
    }
    var checksum = a.substring(0, length)
    while (checksum.length % 2 == 0) {
        checksum = checksum.chunked(2) { if (it[0] == it[1]) 1 else 0 }.joinToString("")
    }
    return checksum.toULong()
}

File: src/main/nim/aoc/2016/day16.nim
import ../aoc

proc checkMe(input: string, length: int): string =
  var a = input
  while a.len() < length:
    a &= '0'
    for i in countdown(a.len() - 2, 0):
      a &= (if a[i] == '0': '1' else: '0')
  
  var checksum = a[0 ..< length]
  while checksum.len() %% 2 == 0:
    var newChecksum = ""
    for i in countup(0, checksum.len() - 1, 2):
      newChecksum &= (if checksum[i] == checksum[i + 1]: '1' else: '0')
    checksum = newChecksum

  checksum

proc day16*(file: string): Solution =
  Solution(
    s1: file.checkMe(272),
    s2: file.checkMe(35651584),
  )
